### PR TITLE
Always pass `ghc-options`

### PR DESCRIPTION
**Pull request by 9999years**
The [documentation for `ghc-shared-options`][1] states that they are
combined with `ghc-options`:

> Additional options for GHC when the package is built as shared library. The options specified via this field are combined with the ones specified via `ghc-options`, and are passed to GHC during both the compile and link phases.

However, _only_ `ghc-shared-options` and not `ghc-options` are passed in many cases.

This is an issue because it requires setting `ghc-shared-options` even if the shared (dynamic) parts of the build don't actually need different options; this has the unpleasant side-effect of causing modules to be compiled twice, effectively doubling compile time! See here, where any non-empty `ghc-shared-options` causes Cabal to not set `-dynamic-too`:

https://github.com/haskell/cabal/blob/acbc0f3a5cc9faf0913ff3e270196693816cec41/Cabal/src/Distribution/Simple/GHC.hs#L1466-L1469

This issue was discovered while integrating the `mold` linker with a Haskell project; when `ghc-shared-options` wasn't set to the same value as `ghc-options`, the `gold` linker would be used instead of `mold` for various links during the build.

[1]: https://cabal.readthedocs.io/en/latest/cabal-package.html#pkg-field-ghc-shared-options

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

I've tested this by using Cabal with this patch applied to build a large commercial Haskell project (the [mercury.com](https://mercury.com) backend) and saw that the `ghc-options` were correctly applied (in this case, that meant observing that no links were performed with `gold` and all links were correctly performed with `mold`, as the options requested, with the exception of the single link described in #4439).

Possibly related issues:

- #4439
- #2715
- #7998

### CODE REVIEW

**Comment by Kleidukos** [comment]
@9999years would you like my help with fixing the conflicts of the PR?

**Comment by 9999years** [comment]
Rebased against latest `master`, PTAL

**Review by fendor — APPROVED** [review]
Essentially, this makes sure that wherever we have shared options, we also pass in the ghc-options. I think the implementatn makes sense. Ideally, we'd have some kind of test for this... But I wouldn't know how.

LGTM!

**Review by geekosaur — APPROVED** [review]
LGTM

File: Cabal/src/Distribution/Simple/GHC.hs
Comment (hsyl20) [inline Cabal/src/Distribution/Simple/GHC.hs:720]: For consistency shouldn't you do the same thing for `ghcOptExtra` of `profOpts` a few line above?

**Comment by ners** [comment]
@9999years thanks for fixing this. Is there anything blocking this PR from getting merged?

I'm keen on using the solution outlined in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9908#note_481299

**Comment by Kleidukos** [comment]
There are some merge conflicts to solve still

**Comment by 9999years** [comment]
@Kleidukos Are you going to actually merge or review it? I've fixed conflicts multiple times with this PR over the last year and it still hasn't moved.

**Comment by Kleidukos** [comment]
@9999years Yes, I thought your reviewers would have merged it. I will take personal care of this PR. Thank you again for your patience. :bow:

**Comment by Kleidukos** [comment]
@9999years would you mind giving a thought to https://github.com/haskell/cabal/pull/8717#discussion_r1227688871 ?

**Comment by Gabriella439** [comment]
Is there a reason that this hasn't been merged?  This PR got two approvals and then it seemed like people wanted it merged and then nothing happened.  Could we get this merged and address any follow-up concerns in a future PR?

**Comment by geekosaur** [comment]
There are still merge conflicts that prevent it from being merged. I think we're still waiting on @Kleidukos but I can take a look at them later today if necessary.

**Comment by Kleidukos** [comment]
> Is there a reason that this hasn't been merged? This PR got two approvals and then it seemed like people wanted it merged and then nothing happened. Could we get this merged and address any follow-up concerns in a future PR?

Maybe GitHub doesn't show you this but this is what I see: 
image

I will try and find the time to rebase this PR this week.

**Comment by Kleidukos** [comment]
@9999years Is there a procedure I can follow to ensure that the patch does what it says?

**Comment by geekosaur** [comment]
@mergify backport 3.12

**Pull request merged**
